ulauncher/api/shared/action/LaunchAppAction.py

```diff
@@ -51,16 +55,12 @@ def run(self):
             else:
                 sanitized_exec = shlex.split(sanitized_exec)
             if hasSystemdRun and not app.get_boolean('X-Ulauncher-Inherit-Scope'):
+                setenv = ["--setenv={}={}".format(k, v) for k, v in env.items()]
                 sanitized_exec = [
                     'systemd-run',
                     '--user',
-                    '--remain-after-exit',
-                ] + sanitized_exec
-
-            env = dict(os.environ.items())
-            # Make sure GDK apps aren't forced to use x11 on wayland due to ulauncher's need to run
-            # under X11 for proper centering.
-            env.pop("GDK_BACKEND", None)
+                    '--remain-after-exit'
+                ] + setenv + sanitized_exec
```

Comment:
It does seems like we are just piling on the callers, and each tweak seems to break a different mode. I wish there was an easy way to write some integration tests for these issues!

Looking at the systemd-run doc, I see the running as a transient scope (using `--scope`) inherits the execution environment, and systemd-run doesn't exit until the command exits. This explains why it worked to pass the env vars through `spawn_async`. However, without the `--scope` it executes in a transient "service", in a fresh environment, which is why the env vars are busted now. So the spawn_async call doesn't need to have the env vars if we use `systemd-run` and if `-E`, but of course they *are* required if a system is configured to use the terminal or doesn't have the systemd-run executable.

Dropping `--scope` means that new transient services are created for each execution, and as far as I can tell, never cleaned up:
```
  run-r34df63b471104419bb191529b3160d6b.service       loaded active exited  /usr/bin/alacritty
  run-r3f7592536bdb48c58c57ba1078ee0317.service       loaded active exited  /usr/bin/gpxsee
  run-r4949f4505d24432294fdfce423283f83.service       loaded active running /usr/bin/alacritty
● run-r4f80efc235104561889d0dfaa59457fb.service       loaded failed failed  /usr/bin/git-dag --prompt
  run-r9e503ce5f19442b38715618f9308d75e.service       loaded active running /usr/bin/alacritty
```

This seems like for a long running system this could get pretty big. Sure we are doing something wrong is there is the "right" way to do this?

Reply:
These were my thoughts exactly, when I was going through the docs. However, what if we use `--scope`, remove `--remain-after-exit` (it doesn't work with `--scope`) and continue using `GLib.spawn_async`. In my limited testing, apps start with the inherited environment as expected and continue to execute even after stopping `ulauncher` itself.

Reply:
@darkfishy Wow. I'm surprised that works! systemd-run is needed specifically for launching apps detached from the main process group when Ulauncher is running as a service with systemd. When I kill ulauncher the apps remain though, so it works I think, and it's less of a convoluted weird solution.

Could you test this also @Jaakkonen? Replacing `--remain-after-exit` with `--scope`? It should work right if the app launches. And I would love some input from @troycurtisjr also before we go with this and push another patch release?

Reply:
Isn't using scope what we changed to address some of the errors people had when not running ulauncher with systemd? https://github.com/Ulauncher/Ulauncher/commit/93c68de2b4c994840e577bb5a3aa5a5659da287b

Is there a way to tell when we are running in our own systemd scope/service? Perhaps we only need to do the systemd-run truck in that case?

Reply:
I reckon the issue #842, which commit 93c68de2b4c994840e577bb5a3aa5a5659da287b addresses, could have been because of resource exhaustion and eventual starvation for new processes, but not necessarily due to the way `ulauncher` was calling `systemd-run`. More specifically, looking at the https://github.com/Ulauncher/Ulauncher/issues/842#issuecomment-963927467 in that issue, one of the errors is that `No PIDs left to attach`.  This could have been due to either system level `pid.max` or some stringent limit on `pid.max` up the `slice` hierarchy getting exhausted.

Without slicing, the new scope/service is placed under the default `app.slice` which may have different resource limits (generally max).

It also seems like `systemd` [encourages](https://systemd.io/DESKTOP_ENVIRONMENTS/#xdg-standardization-for-applications) use of `.service` over `.scope`, though then it means using the `--setenv` hack above.

On a different note, was there a reason behind using `--remain-after-exit`? As I understand, this keeps the transient process loaded and active even after the underlying process has exited.

Reply:
> Isn't using scope what we changed to address some of the errors people had when not running ulauncher with systemd? [93c68de](https://github.com/Ulauncher/Ulauncher/commit/93c68de2b4c994840e577bb5a3aa5a5659da287b)

--scope was only removed because it's incompatible with --remain-after-exit

> Is there a way to tell when we are running in our own systemd scope/service? Perhaps we only need to do the systemd-run truck in that case?

No. Removing --scope is what caused the issue this PR is trying to solve.